extensions/standard-processors/processors/ReplaceText.cpp

```diff
@@ -0,0 +1,359 @@
+          return is_last_line ? applyReplacements(input_line, flow_file) : input_line;
+        case LineByLineEvaluationModeType::EXCEPT_FIRST_LINE:
+          return is_first_line ? input_line : applyReplacements(input_line, flow_file);
+        case LineByLineEvaluationModeType::EXCEPT_LAST_LINE:
+          return is_last_line ? input_line: applyReplacements(input_line, flow_file);
+      }
+      throw Exception{PROCESSOR_EXCEPTION, utils::StringUtils::join_pack("Unsupported ", LineByLineEvaluationMode.getName(), ": ", line_by_line_evaluation_mode_.toString())};
+    }};
+    session->readWrite(flow_file, &read_write_callback);
+    session->transfer(flow_file, Success);
+  } catch (const Exception& exception) {
+    logger_->log_error("Error in ReplaceText (Line-by-Line mode): %s", exception.what());
+    session->transfer(flow_file, Failure);
+  }
+}
+
+std::string ReplaceText::applyReplacements(const std::string& input, const std::shared_ptr<core::FlowFile>& flow_file) const {
+  switch (replacement_strategy_.value()) {
+    case ReplacementStrategyType::PREPEND:
+      return replacement_value_ + input;
+
+    case ReplacementStrategyType::APPEND:
+      return input + replacement_value_;
+
+    case ReplacementStrategyType::REGEX_REPLACE:
+      return applyRegexReplace(input);
+
+    case ReplacementStrategyType::LITERAL_REPLACE:
+      return applyLiteralReplace(input);
+
+    case ReplacementStrategyType::ALWAYS_REPLACE:
+      return replacement_value_;
+
+    case ReplacementStrategyType::SUBSTITUTE_VARIABLES:
+      return applySubstituteVariables(input, flow_file);
+  }
+
+  throw Exception{PROCESSOR_EXCEPTION, utils::StringUtils::join_pack("Unsupported ", ReplacementStrategy.getName(), ": ", replacement_strategy_.toString())};
+}
+
+std::string ReplaceText::applyRegexReplace(const std::string& input) const {
+  const auto [chomped_line, line_ending] = utils::StringUtils::chomp(input);
+  std::string output = std::regex_replace(chomped_line, search_regex_, replacement_value_);
+  return output + line_ending;
+}
+
+std::string ReplaceText::applyLiteralReplace(const std::string& input) const {
+  std::vector<char> output;
+  output.reserve(input.size());
+
+  auto it = input.begin();
+  do {
+    auto found = std::search(it, input.end(), search_value_.begin(), search_value_.end());
+    if (found != input.end()) {
+      std::copy(it, found, std::back_inserter(output));
+      std::copy(replacement_value_.begin(), replacement_value_.end(), std::back_inserter(output));
+      it = found;
+      std::advance(it, search_value_.size());
+    } else {
+      std::copy(it, input.end(), std::back_inserter(output));
+      it = input.end();
+    }
+  } while (it != input.end());
+
+  return std::string{output.begin(), output.end()};
+}
+
+std::string ReplaceText::applySubstituteVariables(const std::string& input, const std::shared_ptr<core::FlowFile>& flow_file) const {
+  static const std::regex PLACEHOLDER{R"(\$\{([^}]+)\})"};
```

Comment:
I see that nifi also uses a hand-rolled variable substitution here, but the attribute substitution already happened when we queried the property and the ExpressionLanguage is enabled, right? it seems to me that standard processors have the EL extension as dependency, so we might be able to rely on that here as well

Reply:
That was my original plan, too, but the expression language substitution logic is hidden inside the expression-language extension, and I didn't think exposing it should be part of this PR.  Also, full expression language support was not a requirement.  We could do this later in a separate PR.